staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -181,6 +181,11 @@ func newAuthenticator(c *cache, config *api.ExecConfig, cluster *clientauthentic
 		connTracker,
 	)
 
+	interactive, err := isInteractive(isTerminalFunc, config)
```

Comment:
Do we want to move this check to take place lazily in `refreshCredsLocked()`? I wonder if someone could hook up a TTY to standard input on a long-running client-go process. Regardless, this seems like an anti-pattern, so I only performed the check once here.

Reply:
Checking it once seems fine to me, though maybe we should check it in `UpdateTransportConfig`?

Reply:
Mmm, I think I just thought of a reason we want this check down in `refreshCredsLocked()`.
1. User types `kubectl create -f -`
2. The kubectl code wires up a `kubernetes.Clientset` which transitively calls `GetAuthenticator()` and `UpdateTransportConfig()` (note: `terminal.IsTerminal()` returns `true` here since standard input is open and a terminal)
3. The kubectl code goes to get the object from standard input
4. User types in their object by hand on the command line
5. User closes standard input with Ctrl-D
6. kubectl receives object from standard input
7. kubectl makes API call with client
8. client goes to refresh exec credentials (note: `terminal.IsTerminal()` returns `false` here since standard input is closed)

Reply:
maybe... but I still think we want something in place that lets `kubectl -f -` mark stdin as unavailable for use by credential plugins to avoid that entirely

I *think* the resource visitor could handle stdin containing two objects:

```
kind: ...
apiVersion: ....
...
---
kind: ...
apiVersion: ....
...
```

by reading the first, making an API call, then reading the second and making another API call

we wouldn't want the credential plugin to consume the second object as stdin

Reply:
mmm. yeah. that makes sense.

here are the current constraints/design goals/opinions that i can think of for this issue:
- applications built on top of client-go should tell client-go whether stdin is available for use by credential plugins
- we should pick a safe default for this behavior so that when an existing application upgrades to client-go v0.22.0 AND that existing application consumes a kubeconfig with a client.authentication.k8s.io/v1 interactive exec plugin, then the interactive behavior will work safely
- it would be really nice if that _safe default for this behavior_ from above would involve client.authentication.k8s.io/v1 interactive exec plugins working out of the box with the existing client-go application that has just upgraded to client-go v0.22.0, since commonly these two components (client-go app and exec plugin) will be developed and distributed by completely separate organizations
- all client-go applications that consume a kubeconfig should care about whether stdin is available for use, because they do not necessarily know what underlying authentication mechanism is in the kubeconfig

we can somewhat easily solve the backend of this flow by adding a new field/parameter to rest config creation that determines whether stdin is safe to use (or we can provide the stdin `os.File` itself?). but i haven't envisioned how we would wire this down from the top of the stack :(

Reply:
> but I still think we want something in place that lets kubectl -f - mark stdin as unavailable for use by credential plugins to avoid that entirely

Alright, I gave this a shot.
- client-go changes: https://github.com/kubernetes/kubernetes/pull/99310/commits/ad7a50bddefcdc1b9f41bc17e38e1c4f3fc45772
- `kubectl` changes: https://github.com/kubernetes/kubernetes/pull/99310/commits/6070cff07e8a3137017e3527e92ee64e80aa5164

What do y'all think? Do we think this is good enough to protect against the "stdin fighting" problem?

Reply:
current approach looks reasonable to me